### ISSUE TITLE
top_file_merging_strategy does not work as documented

### ISSUE THREAD

**Issue by GarthSnyder** [saltstack/salt#34975]
top_file_merging_strategy is documented as a minion-side configuration option. But if you set it to 'same' on the minion side, you'll see that it has no actual effect. 

There's a warning message generated whenever multiple top files are merged. In the current version of 2016.3, the message is "top_file_merging_strategy is set to 'merge' and multiple top files were found. Merging order is not deterministic, it may be desirable to either set top_file_merging_strategy to 'same' or use the 'env_order' configuration parameter to specify the merging order." Older versions have a slightly different message.

Anyway, this warning is always printed whenever you go down the "merge" path in salt/state.py with multiple top files, so it's an easy way to verify that get_tops() is seeing a top_files_merging_strategy of "merge" despite this option being set to 'same' in /etc/salt/minion.

I believe the underlying issue is that __gen_opts() in salt/state.py copies over some option values from the master and does not successfully allow the minion settings to override the master. The code looks roughly like this:

```
def __gen_opts(self, opts):
    ...
    mopts = self.client.master_opts()
    ...
    opts['state_auto_order'] = mopts.get('state_auto_order', opts['state_auto_order'])
    opts['top_file_merging_strategy'] = mopts.get('top_file_merging_strategy', opts.get('top_file_merging_strategy'))
    opts['env_order'] = mopts.get('env_order', opts.get('env_order', []))
    opts['default_top'] = mopts.get('default_top', opts.get('default_top'))
    opts['state_aggregate'] = mopts.get('state_aggregate', opts.get('state_aggregate', False))
    return opts
```

The issue seems to be that any option with a default value will show up in the mopts hash. So whether you set top_file_merging_strategy explicitly or not on the master, it does have a value there, and so the minion's local value always gets clobbered.

I would hazard a guess that some of the other copied options are in this same boat, too.

**Comment by GarthSnyder** [comment]
```
Salt Version:
           Salt: 2016.3.1-147-gf23e8c5
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.12 (default, Jul  5 2016, 01:21:33)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.3
            ZMQ: 4.1.4
 
System Versions:
           dist:   
        machine: amd64
        release: 10.2-RELEASE-p18
         system: FreeBSD
        version: Not Installed

```

**Comment by Ch3LL** [comment]
@GarthSnyder are you referring to running masterless?  when running with `salt-call --local` and the following log added to salt/state.py:

```
 2462             opts['top_file_merging_strategy'] = mopts.get('top_file_merging_strategy',                                                     
 2463                                                           opts.get('top_file_merging_strategy'))                                           
 2464             log.debug('STRATEGY: {0}'.format(opts['top_file_merging_strategy']))             
```

and adding `top_file_merging_strategy` to `same` in the minion config and nothing in the master it logs correctly as `same`.

But when just running `salt-call` without --local it does not work and I still see `merge`. Just want to make sure I understand how you are calling salt when you see this behavior. Thanks.

**Comment by GarthSnyder** [comment]
No, just the normal master/minion configuration with the operation initiated from the master, e.g.

`sudo salt '*' state.show_top`

I'm guessing based on your results that `salt-call --local` adds the option `local_state` to the configuration for the duration of the execution - is that right? There's code at the very top of `__gen_opts() `that looks for this config option and short-circuits the master overrides if it's present:

```
        # If the state is intended to be applied locally, then the local opts
        # should have all of the needed data, otherwise overwrite the local
        # data items with data from the master
        if 'local_state' in opts:
            if opts['local_state']:
                return opts
```

I didn't mention this previously because I couldn't find any other instance of "local_state" in the code base and had figured it was vestigial...

**Comment by GarthSnyder** [comment]
See also [#35045: "top_file_merging_strategy: same" is functionally identical to specifying an environment](https://github.com/saltstack/salt/issues/35045)

**Comment by Ch3LL** [comment]
@GarthSnyder maybe i'm not understanding something very clearly but if you are calling it from the master you can set `top_file_merging_strategy` in the master config. Is there a reason you want it to grab opts from the minion instead?

**Comment by GarthSnyder** [comment]
As the issue title says, the behavior is not as documented; top_file_merging_strategy is presented in the docs as a minion-side option. As far as I can tell, none of the odd overlay behavior is documented anywhere.

Yes, once you read the code and understand what's going on, you can work around it -- is that the recommended solution?

It doesn't really matter since [the feature doesn't work anyway](https://github.com/saltstack/salt/issues/35045) and no one seems to have noticed. Why not just remove it?

**Comment by Ch3LL** [comment]
ahhh gotch ya. Thanks for clarifying. I believer were your confusion came in was from [these](https://docs.saltstack.com/en/latest/ref/configuration/minion.html#top-file-merging-strategy) docs which explain the options you can configure on the minion, which one of them includes `top_file_merging_strategy`. Its not an issue that this option is included on the minion but that its not included in the master configuration options [here](https://docs.saltstack.com/en/latest/ref/configuration/master.html) The reason its not an issue that its in the minion configuration is because you can use a lot of the same options in both the master and minion configuration files if you are for example running in masterless mode. 

I would expect the behavior when calling salt-call and having `top_file_merging_strategy` in the minion that it would actually grab the master opts instead because salt-call actually calls out to the master and should act just like a regular minion. Its in the instance that you are running masterless (`--local) and you are not making a call out to the master that the options from teh minion file should be called. This is currently working when I tested.

So to sum it up we need to add the `top_file_merging_strategy` option to the master configuration docs. Just for thoroughness though there is more information about merging strategies documented [here](https://docs.saltstack.com/en/latest/ref/states/top.html#how-top-files-are-compiled)

Does that suffice for you @GarthSnyder ?

Also, I commented on that other issue concerning the `same` merging strategy not working so we will track that there.

**Comment by GarthSnyder** [comment]
Including `top_file_merging_strategy` in the master-side options would be an incremental improvement, but as long as it's also in the minion-side documentation, I would imagine that people will assume it can be set in either location. And in the typical case, that's not actually true.

The broader issue is that there's a sizable set of minion options that are routinely overridden by master settings, and yet no explicit documentation of this fact appears to exist. Shouldn't the minion-side description of those options say something explicit about this? For example:

"This option can be used only on master servers and standalone minions. When a master server is available, any minion-specified value will be overridden by the corresponding value from the master. This substitution occurs regardless of whether the master has set an explicit value for the option; if it has not, minions will receive the default value from the master side."

**Comment by Ch3LL** [comment]
@GarthSnyder agreed I think as part of this issue report we could add better information as to which options in a minion config are only for masterless.

**Comment by terminalmage** [comment]
@GarthSnyder Just FYI, the ``same`` merging strategy was addressed within the last few days, see [here](https://github.com/saltstack/salt/issues/35045#issuecomment-242854456) for details.

We'll still need to get some of this stuff documented, there is quite a bit in state.py that is able to be set either on the master or minion (with the latter taking precedence and the former establishing a default for all minions), and you're right, it is poorly (or in many cases, not at all) documented what the expected behavior should be.
